[Task] Automated Rule deletion fails if some matched targets cannot be connected

To reproduce, using `smoketest.sh` for consistent setup:

1. Ensure there are no backend stored JMX credentials for the `es.andrewazor.Main` sample applications
2. Create an automated rule definition as follows:
```json
{
  "name": "broken_deletion",
  "description": "",
  "matchExpression": "target.alias == 'es.andrewazor.demo.Main'",
  "eventSpecifier": "template=Continuous,type=TARGET",
  "archivalPeriodSeconds": 0,
  "initialDelaySeconds": 0,
  "preservedArchives": 0,
  "maxAgeSeconds": 0,
  "maxSizeBytes": 0,
  "enabled": true
}
```
3. Notice that this rule should apply to all three 9093/9094/9095 `es.andrewazor.Main` sample applications, The `9093` one does not require JMX credentials, the other two do (and 9095 also requires an SSL cert). Because there are no stored credentials for these targets the connection will fail in the background, so notifications for the rule creation will only appear in the UI for the 9093 target.
4. Delete the new rule definition. Leave the `clean` checkbox checked.
![image](https://user-images.githubusercontent.com/3787464/193856741-dc373586-ab13-43d4-b14b-96421988e190.png)

```
INFO: (10.0.2.100:59538): OPTIONS /api/v2/rules/broken_deletion 204 0ms
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Outgoing WS message: {"meta":{"category":"RuleDeleted","type":{"type":"application","subType":"json"},"serverTime":1664896179},"message":{"name":"broken_deletion","description":"","matchExpression":"target.alias == 'es.andrewazor.demo.Main'","eventSpecifier":"template=Continuous,type=TARGET","archivalPeriodSeconds":0,"initialDelaySeconds":0,"preservedArchives":0,"maxAgeSeconds":0,"maxSizeBytes":0,"enabled":true}}
Hibernate: 
    /* select
        generatedAlias0 
    from
        PluginInfo as generatedAlias0 */ select
            plugininfo0_.id as id1_0_,
            plugininfo0_.callback as callback2_0_,
            plugininfo0_.realm as realm3_0_,
            plugininfo0_.subtree as subtree4_0_ 
        from
            PluginInfo plugininfo0_
Hibernate: 
    /* select
        generatedAlias0 
    from
        StoredCredentials as generatedAlias0 */ select
            storedcred0_.id as id1_1_,
            storedcred0_.matchExpression as matchexp2_1_,
            storedcred0_.password as password3_1_,
            storedcred0_.username as username4_1_ 
        from
            StoredCredentials storedcred0_
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9095/jmxrmi
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: connection attempt failed.

Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Connection for service:jmx:rmi:///jndi/rmi://cryostat:9095/jmxrmi closed
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
io.cryostat.net.web.http.api.v2.ApiException: Authentication failed! Invalid username or password
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.handle(RuleDeleteHandler.java:180)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:121)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:75)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.vertx.ext.web.handler.impl.CorsHandlerImpl.handle(CorsHandlerImpl.java:227)
	at io.vertx.ext.web.handler.impl.CorsHandlerImpl.handle(CorsHandlerImpl.java:41)
	at io.cryostat.net.web.http.generic.CorsEnablingHandler.handle(CorsEnablingHandler.java:113)
	at io.cryostat.net.web.http.generic.CorsEnablingHandler.handle(CorsEnablingHandler.java:57)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:67)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37)
	at io.cryostat.net.web.WebServer.lambda$start$4(WebServer.java:272)
	at io.cryostat.net.HttpServer$HandlerDelegate.handle(HttpServer.java:168)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:57)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:168)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:156)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.openjdk.jmc.rjmx.ConnectionException caused by java.lang.SecurityException: Authentication failed! Invalid username or password
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:441)
	at io.cryostat.core.net.JFRConnection.attemptConnect(JFRConnection.java:300)
	at io.cryostat.core.net.JFRConnection.connect(JFRConnection.java:257)
	at io.cryostat.core.net.JFRConnection.getService(JFRConnection.java:120)
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.lambda$handle$2(RuleDeleteHandler.java:158)
	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:163)
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.handle(RuleDeleteHandler.java:155)
	... 58 more
Caused by: java.lang.SecurityException: Authentication failed! Invalid username or password
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticationFailure(JMXPluggableAuthenticator.java:231)
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticate(JMXPluggableAuthenticator.java:208)
	at jdk.management.agent/sun.management.jmxremote.ConnectorBootstrap$AccessFileCheckerAuthenticator.authenticate(ConnectorBootstrap.java:233)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.doNewClient(RMIServerImpl.java:231)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.newClient(RMIServerImpl.java:198)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:304)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:280)
	at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:165)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl_Stub.newClient(RMIServerImpl_Stub.java:83)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.getConnection(RMIConnector.java:2107)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:321)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connectJmxConnector(RJMXConnection.java:487)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.establishConnection(RJMXConnection.java:472)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:434)
	... 64 more
Caused by: javax.security.auth.login.FailedLoginException: Invalid username or password
	at java.management/com.sun.jmx.remote.security.FileLoginModule.attemptAuthentication(FileLoginModule.java:458)
	at java.management/com.sun.jmx.remote.security.FileLoginModule.login(FileLoginModule.java:309)
	at java.base/javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
	at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:679)
	at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:677)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:677)
	at java.base/javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticate(JMXPluggableAuthenticator.java:195)
	at jdk.management.agent/sun.management.jmxremote.ConnectorBootstrap$AccessFileCheckerAuthenticator.authenticate(ConnectorBootstrap.java:233)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.doNewClient(RMIServerImpl.java:231)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.newClient(RMIServerImpl.java:198)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	... 1 more

Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Removing cached connection for service:jmx:rmi:///jndi/rmi://cryostat:9095/jmxrmi: EXPLICIT
Hibernate: 
    /* select
        generatedAlias0 
    from
        StoredCredentials as generatedAlias0 */ select
            storedcred0_.id as id1_1_,
            storedcred0_.matchExpression as matchexp2_1_,
            storedcred0_.password as password3_1_,
            storedcred0_.username as username4_1_ 
        from
            StoredCredentials storedcred0_
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
Hibernate: 
    /* select
        generatedAlias0 
    from
        StoredCredentials as generatedAlias0 */ select
            storedcred0_.id as id1_1_,
            storedcred0_.matchExpression as matchexp2_1_,
            storedcred0_.password as password3_1_,
            storedcred0_.username as username4_1_ 
        from
            StoredCredentials storedcred0_
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9094/jmxrmi
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: connection attempt failed.

Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Connection for service:jmx:rmi:///jndi/rmi://cryostat:9094/jmxrmi closed
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
io.cryostat.net.web.http.api.v2.ApiException: Authentication failed! Invalid username or password
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.handle(RuleDeleteHandler.java:180)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:121)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:75)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.vertx.ext.web.handler.impl.CorsHandlerImpl.handle(CorsHandlerImpl.java:227)
	at io.vertx.ext.web.handler.impl.CorsHandlerImpl.handle(CorsHandlerImpl.java:41)
	at io.cryostat.net.web.http.generic.CorsEnablingHandler.handle(CorsEnablingHandler.java:113)
	at io.cryostat.net.web.http.generic.CorsEnablingHandler.handle(CorsEnablingHandler.java:57)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:67)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37)
	at io.cryostat.net.web.WebServer.lambda$start$4(WebServer.java:272)
	at io.cryostat.net.HttpServer$HandlerDelegate.handle(HttpServer.java:168)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:57)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:168)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:156)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.openjdk.jmc.rjmx.ConnectionException caused by java.lang.SecurityException: Authentication failed! Invalid username or password
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:441)
	at io.cryostat.core.net.JFRConnection.attemptConnect(JFRConnection.java:300)
	at io.cryostat.core.net.JFRConnection.connect(JFRConnection.java:257)
	at io.cryostat.core.net.JFRConnection.getService(JFRConnection.java:120)
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.lambda$handle$2(RuleDeleteHandler.java:158)
	at io.cryostat.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:163)
	at io.cryostat.net.web.http.api.v2.RuleDeleteHandler.handle(RuleDeleteHandler.java:155)
	... 58 more
Caused by: java.lang.SecurityException: Authentication failed! Invalid username or password
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticationFailure(JMXPluggableAuthenticator.java:231)
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticate(JMXPluggableAuthenticator.java:208)
	at jdk.management.agent/sun.management.jmxremote.ConnectorBootstrap$AccessFileCheckerAuthenticator.authenticate(ConnectorBootstrap.java:233)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.doNewClient(RMIServerImpl.java:231)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.newClient(RMIServerImpl.java:198)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:304)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:280)
	at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:165)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl_Stub.newClient(RMIServerImpl_Stub.java:83)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.getConnection(RMIConnector.java:2107)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:321)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connectJmxConnector(RJMXConnection.java:487)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.establishConnection(RJMXConnection.java:464)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:434)
	... 64 more
Caused by: javax.security.auth.login.FailedLoginException: Invalid username or password
	at java.management/com.sun.jmx.remote.security.FileLoginModule.attemptAuthentication(FileLoginModule.java:458)
	at java.management/com.sun.jmx.remote.security.FileLoginModule.login(FileLoginModule.java:309)
	at java.base/javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
	at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:679)
	at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:677)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base/javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:677)
	at java.base/javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at java.management/com.sun.jmx.remote.security.JMXPluggableAuthenticator.authenticate(JMXPluggableAuthenticator.java:195)
	at jdk.management.agent/sun.management.jmxremote.ConnectorBootstrap$AccessFileCheckerAuthenticator.authenticate(ConnectorBootstrap.java:233)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.doNewClient(RMIServerImpl.java:231)
	at java.management.rmi/javax.management.remote.rmi.RMIServerImpl.newClient(RMIServerImpl.java:198)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	... 1 more

Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: Removing cached connection for service:jmx:rmi:///jndi/rmi://cryostat:9094/jmxrmi: EXPLICIT
Oct 04, 2022 3:09:39 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:59542): DELETE /api/v2/rules/broken_deletion 500 69ms
Oct 04, 2022 3:09:49 PM io.cryostat.core.log.Logger info
INFO: Removing cached connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi: EXPIRED
Oct 04, 2022 3:09:49 PM io.cryostat.core.log.Logger info
INFO: Connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi closed
```

The request fails because the 9094/9095 targets cannot be connected to attempt the rule cleanup. This should be silently ignored (just logged) and the deletion request should have a `2xx` response code.

--- a/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
@@ -63,7 +63,7 @@ import com.google.gson.Gson;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import io.vertx.core.http.HttpMethod;
 
-class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.CleanupFailure>> {
+class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
 
     private static final String DELETE_RULE_CATEGORY = "RuleDeleted";
     static final String PATH = RuleGetHandler.PATH;
@@ -126,7 +126,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.
     }
 
     @Override
-    public IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> handle(
+    public IntermediateResponse<Void> handle(
             RequestParameters params) throws ApiException {
         String name = params.getPathParams().get(Rule.Attribute.NAME.getSerialKey());
         if (!ruleRegistry.hasRuleByName(name)) {
@@ -145,7 +145,6 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.
                 .message(rule)
                 .build()
                 .send();
-        List<CleanupFailure> failures = new ArrayList<>();
         if (Boolean.valueOf(params.getQueryParams().get(CLEAN_PARAM))) {
             for (ServiceRef ref : storage.listDiscoverableServices()) {
                 if (!ruleRegistry.applies(rule, ref)) {
@@ -166,31 +165,16 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.
                                                     try {
                                                         conn.getService().stop(r);
                                                     } catch (Exception e) {
-                                                        logger.error(new ApiException(500, e));
-                                                        CleanupFailure failure =
-                                                                new CleanupFailure();
-                                                        failure.ref = ref;
-                                                        failure.message = e.getMessage();
-                                                        failures.add(failure);
+                                                        logger.error(e);
                                                     }
                                                 });
                                 return null;
                             });
                 } catch (Exception e) {
                     logger.error(e);
-                    CleanupFailure failure = new CleanupFailure();
-                    failure.ref = ref;
-                    failure.message = e.getMessage();
-                    failures.add(failure);
                 }
             }
         }
-        return new IntermediateResponse<List<CleanupFailure>>().body(null);
-    }
-
-    @SuppressFBWarnings("URF_UNREAD_FIELD")
-    static class CleanupFailure {
-        ServiceRef ref;
-        String message;
+        return new IntermediateResponse<Void>().body(null);
     }
 }

--- a/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
@@ -177,7 +177,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.
                                 return null;
                             });
                 } catch (Exception e) {
-                    logger.error(new ApiException(500, e));
+                    logger.error(e);
                     CleanupFailure failure = new CleanupFailure();
                     failure.ref = ref;
                     failure.message = e.getMessage();
@@ -185,11 +185,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<List<RuleDeleteHandler.
                 }
             }
         }
-        if (failures.size() == 0) {
-            return new IntermediateResponse<List<CleanupFailure>>().body(null);
-        } else {
-            return new IntermediateResponse<List<CleanupFailure>>().statusCode(500).body(failures);
-        }
+        return new IntermediateResponse<List<CleanupFailure>>().body(null);
     }
 
     @SuppressFBWarnings("URF_UNREAD_FIELD")

--- a/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
@@ -50,11 +50,11 @@ import io.cryostat.discovery.DiscoveryStorage;
 import io.cryostat.messaging.notifications.NotificationFactory;
 import io.cryostat.net.AuthManager;
 import io.cryostat.net.ConnectionDescriptor;
-import io.cryostat.net.TargetConnectionManager;
 import io.cryostat.net.security.ResourceAction;
 import io.cryostat.net.web.http.HttpMimeType;
 import io.cryostat.net.web.http.api.ApiVersion;
 import io.cryostat.platform.ServiceRef;
+import io.cryostat.recordings.RecordingTargetHelper;
 import io.cryostat.rules.Rule;
 import io.cryostat.rules.RuleRegistry;
 
@@ -70,7 +70,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
 
     private final Vertx vertx;
     private final RuleRegistry ruleRegistry;
-    private final TargetConnectionManager targetConnectionManager;
+    private final RecordingTargetHelper recordings;
     private final DiscoveryStorage storage;
     private final CredentialsManager credentialsManager;
     private final NotificationFactory notificationFactory;
@@ -81,7 +81,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
             Vertx vertx,
             AuthManager auth,
             RuleRegistry ruleRegistry,
-            TargetConnectionManager targetConnectionManager,
+            RecordingTargetHelper recordings,
             DiscoveryStorage storage,
             CredentialsManager credentialsManager,
             NotificationFactory notificationFactory,
@@ -90,7 +90,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
         super(auth, gson);
         this.vertx = vertx;
         this.ruleRegistry = ruleRegistry;
-        this.targetConnectionManager = targetConnectionManager;
+        this.recordings = recordings;
         this.storage = storage;
         this.credentialsManager = credentialsManager;
         this.notificationFactory = notificationFactory;
@@ -147,61 +147,34 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
                 .build()
                 .send();
         if (Boolean.valueOf(params.getQueryParams().get(CLEAN_PARAM))) {
-            vertx.executeBlocking(promise -> {
-                cleanup(rule);
-                promise.complete();
-            });
+            vertx.executeBlocking(
+                    promise -> {
+                        try {
+                            cleanup(rule);
+                            promise.complete();
+                        } catch (Exception e) {
+                            promise.fail(e);
+                        }
+                    });
         }
         return new IntermediateResponse<Void>().body(null);
     }
 
     private void cleanup(Rule rule) {
         for (ServiceRef ref : storage.listDiscoverableServices()) {
-            vertx.<Boolean>executeBlocking(
+            vertx.executeBlocking(
                     promise -> {
                         try {
-                            promise.complete(ruleRegistry.applies(rule, ref));
-                            if (!ruleRegistry.applies(rule, ref)) {
-                                promise.complete(null);
+                            if (ruleRegistry.applies(rule, ref)) {
+                                ConnectionDescriptor cd =
+                                        new ConnectionDescriptor(
+                                                ref, credentialsManager.getCredentials(ref));
+                                recordings.stopRecording(cd, rule.getRecordingName());
                             }
-                        } catch (Exception e) {
-                            promise.fail(e);
-                        }
-                    },
-                    false,
-                    result -> {
-                        if (result.failed()) {
-                            logger.error(new RuntimeException(result.cause()));
-                            return;
-                        }
-                        if (!result.result()) {
-                            return;
-                        }
-                        try {
-                            targetConnectionManager.executeConnectedTaskAsync(
-                                    new ConnectionDescriptor(
-                                            ref, credentialsManager.getCredentials(ref)),
-                                    conn -> {
-                                        conn.getService().getAvailableRecordings().stream()
-                                                .filter(
-                                                        rec ->
-                                                                rec.getName()
-                                                                        .equals(
-                                                                                rule
-                                                                                        .getRecordingName()))
-                                                .findFirst()
-                                                .ifPresent(
-                                                        r -> {
-                                                            try {
-                                                                conn.getService().stop(r);
-                                                            } catch (Exception e) {
-                                                                logger.error(e);
-                                                            }
-                                                        });
-                                        return null;
-                                    });
+                            promise.complete();
                         } catch (Exception e) {
                             logger.error(e);
+                            promise.fail(e);
                         }
                     });
         }

--- a/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java
@@ -38,7 +38,6 @@
 package io.cryostat.net.web.http.api.v2;
 
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
 import java.util.Set;
@@ -60,7 +59,7 @@ import io.cryostat.rules.Rule;
 import io.cryostat.rules.RuleRegistry;
 
 import com.google.gson.Gson;
-import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.vertx.core.Vertx;
 import io.vertx.core.http.HttpMethod;
 
 class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
@@ -69,6 +68,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
     static final String PATH = RuleGetHandler.PATH;
     static final String CLEAN_PARAM = "clean";
 
+    private final Vertx vertx;
     private final RuleRegistry ruleRegistry;
     private final TargetConnectionManager targetConnectionManager;
     private final DiscoveryStorage storage;
@@ -78,6 +78,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
 
     @Inject
     RuleDeleteHandler(
+            Vertx vertx,
             AuthManager auth,
             RuleRegistry ruleRegistry,
             TargetConnectionManager targetConnectionManager,
@@ -87,6 +88,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
             Gson gson,
             Logger logger) {
         super(auth, gson);
+        this.vertx = vertx;
         this.ruleRegistry = ruleRegistry;
         this.targetConnectionManager = targetConnectionManager;
         this.storage = storage;
@@ -126,8 +128,7 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
     }
 
     @Override
-    public IntermediateResponse<Void> handle(
-            RequestParameters params) throws ApiException {
+    public IntermediateResponse<Void> handle(RequestParameters params) throws ApiException {
         String name = params.getPathParams().get(Rule.Attribute.NAME.getSerialKey());
         if (!ruleRegistry.hasRuleByName(name)) {
             throw new ApiException(404);
@@ -146,35 +147,63 @@ class RuleDeleteHandler extends AbstractV2RequestHandler<Void> {
                 .build()
                 .send();
         if (Boolean.valueOf(params.getQueryParams().get(CLEAN_PARAM))) {
-            for (ServiceRef ref : storage.listDiscoverableServices()) {
-                if (!ruleRegistry.applies(rule, ref)) {
-                    continue;
-                }
-                try {
-                    targetConnectionManager.executeConnectedTask(
-                            new ConnectionDescriptor(ref, credentialsManager.getCredentials(ref)),
-                            conn -> {
-                                conn.getService().getAvailableRecordings().stream()
-                                        .filter(
-                                                rec ->
-                                                        rec.getName()
-                                                                .equals(rule.getRecordingName()))
-                                        .findFirst()
-                                        .ifPresent(
-                                                r -> {
-                                                    try {
-                                                        conn.getService().stop(r);
-                                                    } catch (Exception e) {
-                                                        logger.error(e);
-                                                    }
-                                                });
-                                return null;
-                            });
-                } catch (Exception e) {
-                    logger.error(e);
-                }
-            }
+            vertx.executeBlocking(promise -> {
+                cleanup(rule);
+                promise.complete();
+            });
         }
         return new IntermediateResponse<Void>().body(null);
+    }
+
+    private void cleanup(Rule rule) {
+        for (ServiceRef ref : storage.listDiscoverableServices()) {
+            vertx.<Boolean>executeBlocking(
+                    promise -> {
+                        try {
+                            promise.complete(ruleRegistry.applies(rule, ref));
+                            if (!ruleRegistry.applies(rule, ref)) {
+                                promise.complete(null);
+                            }
+                        } catch (Exception e) {
+                            promise.fail(e);
+                        }
+                    },
+                    false,
+                    result -> {
+                        if (result.failed()) {
+                            logger.error(new RuntimeException(result.cause()));
+                            return;
+                        }
+                        if (!result.result()) {
+                            return;
+                        }
+                        try {
+                            targetConnectionManager.executeConnectedTaskAsync(
+                                    new ConnectionDescriptor(
+                                            ref, credentialsManager.getCredentials(ref)),
+                                    conn -> {
+                                        conn.getService().getAvailableRecordings().stream()
+                                                .filter(
+                                                        rec ->
+                                                                rec.getName()
+                                                                        .equals(
+                                                                                rule
+                                                                                        .getRecordingName()))
+                                                .findFirst()
+                                                .ifPresent(
+                                                        r -> {
+                                                            try {
+                                                                conn.getService().stop(r);
+                                                            } catch (Exception e) {
+                                                                logger.error(e);
+                                                            }
+                                                        });
+                                        return null;
+                                    });
+                        } catch (Exception e) {
+                            logger.error(e);
+                        }
+                    });
+        }
     }
 }

--- a/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
+++ b/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
@@ -47,6 +47,7 @@ import org.openjdk.jmc.rjmx.services.jfr.IFlightRecorderService;
 import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
 
 import io.cryostat.MainModule;
+import io.cryostat.MockVertx;
 import io.cryostat.configuration.CredentialsManager;
 import io.cryostat.core.FlightRecorderException;
 import io.cryostat.core.log.Logger;
@@ -65,6 +66,7 @@ import io.cryostat.rules.RuleRegistry;
 
 import com.google.gson.Gson;
 import io.vertx.core.MultiMap;
+import io.vertx.core.Vertx;
 import io.vertx.core.http.HttpMethod;
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.Matchers;
@@ -81,6 +83,7 @@ import org.mockito.junit.jupiter.MockitoExtension;
 class RuleDeleteHandlerTest {
 
     RuleDeleteHandler handler;
+    Vertx vertx = MockVertx.vertx();
     @Mock AuthManager auth;
     @Mock RuleRegistry registry;
     @Mock TargetConnectionManager targetConnectionManager;
@@ -110,6 +113,7 @@ class RuleDeleteHandlerTest {
         Mockito.lenient().when(notificationBuilder.build()).thenReturn(notification);
         this.handler =
                 new RuleDeleteHandler(
+                        vertx,
                         auth,
                         registry,
                         targetConnectionManager,

--- a/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
+++ b/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
@@ -185,8 +185,7 @@ class RuleDeleteHandlerTest {
             Mockito.when(registry.hasRuleByName(testRuleName)).thenReturn(true);
             Mockito.when(registry.getRule(testRuleName)).thenReturn(Optional.of(rule));
 
-            IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> response =
-                    handler.handle(params);
+            IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
             Mockito.verify(notificationFactory).createBuilder();
@@ -206,7 +205,7 @@ class RuleDeleteHandlerTest {
         }
 
         @Test
-        void shouldRespondWith500ForCleanupFailures() throws Exception {
+        void shouldRespondWith200ForCleanupFailures() throws Exception {
             Mockito.when(params.getPathParams()).thenReturn(Map.of("name", testRuleName));
             MultiMap queryParams = MultiMap.caseInsensitiveMultiMap();
             queryParams.set("clean", "true");
@@ -233,18 +232,10 @@ class RuleDeleteHandlerTest {
             Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
                     .thenThrow(exception);
 
-            IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> response =
-                    handler.handle(params);
-            MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(500));
+            IntermediateResponse<Void> response = handler.handle(params);
+            MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
-            RuleDeleteHandler.CleanupFailure failure = new RuleDeleteHandler.CleanupFailure();
-            failure.ref = serviceRef;
-            failure.message = exception.getMessage();
-            List<RuleDeleteHandler.CleanupFailure> actualList = response.getBody();
-            MatcherAssert.assertThat(actualList, Matchers.hasSize(1));
-            RuleDeleteHandler.CleanupFailure actual = actualList.get(0);
-            MatcherAssert.assertThat(actual.ref, Matchers.sameInstance(serviceRef));
-            MatcherAssert.assertThat(actual.message, Matchers.equalTo(exception.getMessage()));
+            Mockito.verify(logger).error(exception);
         }
 
         @Test
@@ -283,8 +274,7 @@ class RuleDeleteHandlerTest {
 
             Mockito.when(service.getAvailableRecordings()).thenReturn(List.of());
 
-            IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> response =
-                    handler.handle(params);
+            IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
         }
 
@@ -326,8 +316,7 @@ class RuleDeleteHandlerTest {
             Mockito.when(service.getAvailableRecordings()).thenReturn(List.of(recording));
             Mockito.when(recording.getName()).thenReturn(rule.getRecordingName());
 
-            IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> response =
-                    handler.handle(params);
+            IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
             Mockito.verify(service).stop(recording);
@@ -335,7 +324,7 @@ class RuleDeleteHandlerTest {
         }
 
         @Test
-        void shouldRespondWith500AfterUnsuccessfulCleanup() throws Exception {
+        void shouldRespondWith200AfterUnsuccessfulCleanup() throws Exception {
             Mockito.when(params.getPathParams()).thenReturn(Map.of("name", testRuleName));
             MultiMap queryParams = MultiMap.caseInsensitiveMultiMap();
             queryParams.set("clean", "true");
@@ -376,21 +365,10 @@ class RuleDeleteHandlerTest {
             Mockito.when(service.getAvailableRecordings()).thenReturn(List.of(recording));
             Mockito.when(recording.getName()).thenReturn(rule.getRecordingName());
 
-            IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>> response =
-                    handler.handle(params);
-            MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(500));
+            IntermediateResponse<Void> response = handler.handle(params);
+            MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
-            RuleDeleteHandler.CleanupFailure failure = new RuleDeleteHandler.CleanupFailure();
-            failure.ref = serviceRef;
-            failure.message = exception.getMessage();
-            List<RuleDeleteHandler.CleanupFailure> actualList = response.getBody();
-            MatcherAssert.assertThat(actualList, Matchers.hasSize(1));
-            RuleDeleteHandler.CleanupFailure actual = actualList.get(0);
-            MatcherAssert.assertThat(actual.ref, Matchers.sameInstance(serviceRef));
-            MatcherAssert.assertThat(actual.message, Matchers.equalTo(exception.getMessage()));
-
-            Mockito.verify(service).stop(recording);
-            Mockito.verify(service, Mockito.never()).close(recording);
+            Mockito.verify(logger).error(exception);
         }
     }
 }

--- a/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
+++ b/src/test/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandlerTest.java
@@ -43,24 +43,20 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 
-import org.openjdk.jmc.rjmx.services.jfr.IFlightRecorderService;
-import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
-
 import io.cryostat.MainModule;
 import io.cryostat.MockVertx;
 import io.cryostat.configuration.CredentialsManager;
 import io.cryostat.core.FlightRecorderException;
 import io.cryostat.core.log.Logger;
-import io.cryostat.core.net.JFRConnection;
 import io.cryostat.discovery.DiscoveryStorage;
 import io.cryostat.messaging.notifications.Notification;
 import io.cryostat.messaging.notifications.NotificationFactory;
 import io.cryostat.net.AuthManager;
-import io.cryostat.net.TargetConnectionManager;
 import io.cryostat.net.security.ResourceAction;
 import io.cryostat.net.web.http.HttpMimeType;
 import io.cryostat.net.web.http.api.ApiVersion;
 import io.cryostat.platform.ServiceRef;
+import io.cryostat.recordings.RecordingTargetHelper;
 import io.cryostat.rules.Rule;
 import io.cryostat.rules.RuleRegistry;
 
@@ -86,7 +82,7 @@ class RuleDeleteHandlerTest {
     Vertx vertx = MockVertx.vertx();
     @Mock AuthManager auth;
     @Mock RuleRegistry registry;
-    @Mock TargetConnectionManager targetConnectionManager;
+    @Mock RecordingTargetHelper recordingTargetHelper;
     @Mock DiscoveryStorage storage;
     @Mock CredentialsManager credentialsManager;
     @Mock NotificationFactory notificationFactory;
@@ -116,7 +112,7 @@ class RuleDeleteHandlerTest {
                         vertx,
                         auth,
                         registry,
-                        targetConnectionManager,
+                        recordingTargetHelper,
                         storage,
                         credentialsManager,
                         notificationFactory,
@@ -206,6 +202,13 @@ class RuleDeleteHandlerTest {
             ApiException ex =
                     Assertions.assertThrows(ApiException.class, () -> handler.handle(params));
             MatcherAssert.assertThat(ex.getStatusCode(), Matchers.equalTo(404));
+
+            Mockito.verify(vertx, Mockito.never()).executeBlocking(Mockito.any());
+            Mockito.verify(registry, Mockito.never()).deleteRule(Mockito.any(Rule.class));
+            Mockito.verify(registry, Mockito.never()).deleteRule(Mockito.anyString());
+            Mockito.verify(registry, Mockito.never()).applies(Mockito.any(), Mockito.any());
+            Mockito.verify(recordingTargetHelper, Mockito.never())
+                    .stopRecording(Mockito.any(), Mockito.any());
         }
 
         @Test
@@ -233,12 +236,17 @@ class RuleDeleteHandlerTest {
 
             FlightRecorderException exception =
                     new FlightRecorderException(new Exception("test message"));
-            Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
+            Mockito.when(recordingTargetHelper.stopRecording(Mockito.any(), Mockito.any()))
                     .thenThrow(exception);
 
             IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
+            Mockito.verify(vertx, Mockito.times(2)).executeBlocking(Mockito.any());
+            Mockito.verify(registry).deleteRule(rule);
+            Mockito.verify(registry).applies(rule, serviceRef);
+            Mockito.verify(recordingTargetHelper)
+                    .stopRecording(Mockito.any(), Mockito.eq(rule.getRecordingName()));
             Mockito.verify(logger).error(exception);
         }
 
@@ -257,29 +265,17 @@ class RuleDeleteHandlerTest {
                             .build();
             Mockito.when(registry.hasRuleByName(testRuleName)).thenReturn(true);
             Mockito.when(registry.getRule(testRuleName)).thenReturn(Optional.of(rule));
-            Mockito.when(registry.applies(Mockito.any(), Mockito.any())).thenReturn(true);
 
-            ServiceRef serviceRef =
-                    new ServiceRef(
-                            new URI("service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi"),
-                            "io.cryostat.Cryostat");
-            Mockito.when(storage.listDiscoverableServices()).thenReturn(List.of(serviceRef));
-
-            JFRConnection connection = Mockito.mock(JFRConnection.class);
-            Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
-                    .thenAnswer(
-                            arg0 ->
-                                    ((TargetConnectionManager.ConnectedTask<Object>)
-                                                    arg0.getArgument(1))
-                                            .execute(connection));
-
-            IFlightRecorderService service = Mockito.mock(IFlightRecorderService.class);
-            Mockito.when(connection.getService()).thenReturn(service);
-
-            Mockito.when(service.getAvailableRecordings()).thenReturn(List.of());
+            Mockito.when(storage.listDiscoverableServices()).thenReturn(List.of());
 
             IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
+
+            Mockito.verify(vertx, Mockito.times(1)).executeBlocking(Mockito.any());
+            Mockito.verify(registry).deleteRule(rule);
+            Mockito.verify(registry, Mockito.never()).applies(Mockito.any(), Mockito.any());
+            Mockito.verify(recordingTargetHelper, Mockito.never())
+                    .stopRecording(Mockito.any(), Mockito.eq(rule.getRecordingName()));
         }
 
         @Test
@@ -305,74 +301,14 @@ class RuleDeleteHandlerTest {
                             "io.cryostat.Cryostat");
             Mockito.when(storage.listDiscoverableServices()).thenReturn(List.of(serviceRef));
 
-            JFRConnection connection = Mockito.mock(JFRConnection.class);
-            Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
-                    .thenAnswer(
-                            arg0 ->
-                                    ((TargetConnectionManager.ConnectedTask<Object>)
-                                                    arg0.getArgument(1))
-                                            .execute(connection));
-
-            IFlightRecorderService service = Mockito.mock(IFlightRecorderService.class);
-            Mockito.when(connection.getService()).thenReturn(service);
-
-            IRecordingDescriptor recording = Mockito.mock(IRecordingDescriptor.class);
-            Mockito.when(service.getAvailableRecordings()).thenReturn(List.of(recording));
-            Mockito.when(recording.getName()).thenReturn(rule.getRecordingName());
-
             IntermediateResponse<Void> response = handler.handle(params);
             MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
 
-            Mockito.verify(service).stop(recording);
-            Mockito.verify(service, Mockito.never()).close(recording);
-        }
-
-        @Test
-        void shouldRespondWith200AfterUnsuccessfulCleanup() throws Exception {
-            Mockito.when(params.getPathParams()).thenReturn(Map.of("name", testRuleName));
-            MultiMap queryParams = MultiMap.caseInsensitiveMultiMap();
-            queryParams.set("clean", "true");
-            Mockito.when(params.getQueryParams()).thenReturn(queryParams);
-
-            Rule rule =
-                    new Rule.Builder()
-                            .name(testRuleName)
-                            .matchExpression("true")
-                            .eventSpecifier("template=Continuous")
-                            .build();
-            Mockito.when(registry.hasRuleByName(testRuleName)).thenReturn(true);
-            Mockito.when(registry.getRule(testRuleName)).thenReturn(Optional.of(rule));
-            Mockito.when(registry.applies(Mockito.any(), Mockito.any())).thenReturn(true);
-
-            ServiceRef serviceRef =
-                    new ServiceRef(
-                            new URI("service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi"),
-                            "io.cryostat.Cryostat");
-            Mockito.when(storage.listDiscoverableServices()).thenReturn(List.of(serviceRef));
-
-            JFRConnection connection = Mockito.mock(JFRConnection.class);
-            Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
-                    .thenAnswer(
-                            arg0 ->
-                                    ((TargetConnectionManager.ConnectedTask<Object>)
-                                                    arg0.getArgument(1))
-                                            .execute(connection));
-
-            IFlightRecorderService service = Mockito.mock(IFlightRecorderService.class);
-            Mockito.when(connection.getService()).thenReturn(service);
-
-            org.openjdk.jmc.rjmx.services.jfr.FlightRecorderException exception =
-                    new org.openjdk.jmc.rjmx.services.jfr.FlightRecorderException("test message");
-            Mockito.doThrow(exception).when(service).stop(Mockito.any());
-
-            IRecordingDescriptor recording = Mockito.mock(IRecordingDescriptor.class);
-            Mockito.when(service.getAvailableRecordings()).thenReturn(List.of(recording));
-            Mockito.when(recording.getName()).thenReturn(rule.getRecordingName());
-
-            IntermediateResponse<Void> response = handler.handle(params);
-            MatcherAssert.assertThat(response.getStatusCode(), Matchers.equalTo(200));
-
-            Mockito.verify(logger).error(exception);
+            Mockito.verify(vertx, Mockito.times(2)).executeBlocking(Mockito.any());
+            Mockito.verify(registry).deleteRule(rule);
+            Mockito.verify(registry).applies(Mockito.any(), Mockito.any());
+            Mockito.verify(recordingTargetHelper)
+                    .stopRecording(Mockito.any(), Mockito.eq(rule.getRecordingName()));
         }
     }
 }

